File: 001-Fibonacci/cpp/fibonacci_memoized.cpp
#include<bits/stdc++.h>
using namespace std;

map<int, int> memo;

int fibonacci(int n) {
    if (n == 0 || n == 1) {
        return 1;
    }
    else if (memo.find(n) != memo.end()) {
        return memo[n];
    }
    else {
        memo[n] = fibonacci(n - 1) + fibonacci(n - 2);
        return memo[n];
    }
}

int main() {
    int n = 6;
    cout << "Calculating fibonacci of " << n << " using memo" << endl;
    cout << fibonacci(n) << endl;
    return 0;
}

File: 001-Fibonacci/python/fibonacci_memoized.py
memo = {}

def fibonacci(n):
    if n == 0 or n == 1:
        return 1
    elif n in memo:
        return memo[n]
    else:
        memo[n] = fibonacci(n - 1) + fibonacci(n - 2)
        return memo[n]

if __name__ == '__main__':
    n = 6
    print("Calculating fibonacci of {} using memo".format(n))
    print(fibonacci(n))
